WIP: adding tilings to outer loops fusions

abialas1 opened the pull request:
Do not merge. Works only on simplest cases

flaub commented on pmlc/dialect/pxa/transforms/fusion.cc:
```diff
@@ -309,10 +338,14 @@ struct FusionInfo {
     for (auto &raw : readAfterWrites) {
-      considerPlan(raw.first, raw.second);
+      auto write = cast<AffineWriteOpInterface>(raw.first);
+      auto read = cast<AffineReadOpInterface>(raw.second);
```
PXA doesn't actually implement these interfaces. We should probably create our own PXA based interfaces, and we've talked about doing so, but it hasn't been implemented yet.

abialas1 replied:
I saw that Affine interface was used for load and vector_load ops, so added the same to reduce. Let me separate these two based on the Affine ones.

jbruestle commented on pmlc/dialect/pxa/transforms/fusion.cc:
```diff
@@ -16,8 +18,8 @@ namespace pmlc::dialect::pxa {
 
-using WriteRead = std::pair<PxaReduceOp, PxaLoadOp>;
-using WriteWrite = std::pair<PxaReduceOp, PxaReduceOp>;
+using WriteRead = std::pair<Operation *, Operation *>;
```
maybe use PxaWriteOpInterface, PxaReadOpInterface here?

jbruestle commented on pmlc/dialect/pxa/ir/interfaces.td:
```diff
@@ -0,0 +1,169 @@
+  ];
+}
+
+def PxaWriteOpInterface : PxaOpInterface<"PxaWriteOpInterface"> {
```
Should we call this PxaReduceOpInterface?  Write doesn't imply the fact that a reduction is happening (unless type is assign).

jbruestle commented on pmlc/dialect/pxa/transforms/tile.cc:
```diff
@@ -64,4 +64,55 @@ AffineParallelOp performTiling(AffineParallelOp op,
   return inner;
 }
 
+AffineParallelOp undoTiling(AffineParallelOp op,
```
This code looks correct to me, but a small test of tile / untile would be good.

jbruestle commented on pmlc/dialect/pxa/transforms/fusion.cc:
```diff
@@ -189,6 +210,14 @@ struct FusionInfo {
+    // Add tiling if needed
+    if (needsTiling) {
+      if (reverseFusion)
+        performTiling(aInfo.op, tileSizes);
```
I presume that the code for untiling isn't yet written?  I realize this is a WIP.

abialas1 replied:
yes, please see the latest checkin which has the tiling\untiling implemented

jbruestle reviewed (commented):
Directionally, looks very good, I like the interface refactors and the untile method seems like it was straightforward.  Some minor thoughts inline.

abialas1 commented on pmlc/conversion/gpu/kernel_outlining.cc:
```diff
@@ -157,13 +157,17 @@ class GpuKernelOutliningPass
       auto triple = spirv::VerCapExtAttr::get(
-          spirv::Version::V_1_0,
-          {spirv::Capability::Shader, spirv::Capability::Groups,
+          spirv::Version::V_1_5,
```
yes, actually they are in the separate PR already, I need these for local testing, will clean it up before moving from WIP to production

abialas1 commented:
The PR got huge and would like to push it at this point to not struggle with the code rebases. The functionality should be transparent to what we have on plaidml-v1 branch right now, and added more options for fusing, added vector support for dataflow_opt, I1StorageToI32 and other changes needed for fusings. I will put detailed status for fusing in https://github.com/plaidml/plaidml/issues/1381

jbruestle commented:
Totally agree, In general I'm happy land code even if it's not complete as long as it doesn't break anything and has gotten code review.  However right now this code is failing CI hard, so that would need to be fixed first.  Also, it would be slightly nicer to land this in 2-3 chunks since there are a few unrelated changes, but it would also be nice to land it soon, so your call on whether or not to split it up.

abialas1 commented:
Hi Jeremy, dividing this code into several chunks is doable, however will need some work on my side to have it separate, as some chunks are dependent on others to have the full pipeline working. Unless it is really needed I would like to focus on perf analysis instead. Let me know if that would be ok.
Basically there are 4 parts to this checkin:
1. Interfaces for PXA Read and Reduce
2. Vector support for I1ToI32Storage
3. Vector support for dataFlowOpt
4. Parametrized tiled fusions + undoTiling

jbruestle commented:
I think in the interest of time, it's OK to land this basically as is.  However in CI it seems to be breaking one test.  You should be able to repro via:

bazelisk run --//plaidml:device=vulkan.0 --//plaidml:target=intel_gen //plaidml/edsl/tests:cc_test

It's probably an actual edge case that needs proper handling, so once you root cause it, maybe add a simplified lit test if it's not too much work.  Also, I'm a bit worried about architecture specific stuff (subgroups) leaking too much into general purpose stuff (fusion), however I think I have a way to fix it in this case (basically compute the tiling to use based on the the relations of the loop indexes rather than the attribute), and I can refactor after this lands, so that's not a blocker.
